### DISCUSSION TITLE
Why is `GITHUB_OUTPUT` more secure than `::set-output`?

### DISCUSSION THREAD

**Discussion by altaurog** [community/community#40193] [Actions] [answered]
### Select Topic Area

Question

### Body

[As announced](https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/), the mechanism for setting output values from a github actions step is changing.  According to the announcement, this is to prevent “untrusted logged data” from using `save-state` and `set-output`.

I must be missing something.  Why is `GITHUB_OUTPUT` more secure?  What untrusted output would be able to emit `::set-output` but not have access to the `GITHUB_OUTPUT` environment variable (and corresponding file on disk)?

**Comment by airtower-luna** [comment, marked as answer]
I assume it's the same issue [as with `set-env`, which was deprecated over 2 years ago](https://github.blog/changelog/2020-10-01-github-actions-deprecating-set-env-and-add-path-commands/): It's *easy* for untrusted data to end up on stdout, and then it could modify the environment, and do lots of evil stuff. Same applies to outputs, though outputs are narrower in where they are used.

It's much harder to trick a workflow into writing to a dedicated file (especially if its name isn't predictable). Mostly you need to make sure to sanitize any untrusted input before writing it to the outputs (or environment) file.

Depending on where you use untrusted input there could be other issues though, see: https://securitylab.github.com/research/github-actions-untrusted-input/
